WIP: runtime changes for static interface method support

trylek opened the pull request:
Apparently I still don't have push rights to your fork so for now I'm pushing the commits to my fork. I originally tried to create this PR against your version of the branch but GitHub doesn't let me select it as the base repo, probably due to the fact that I'm not authorized to push. I assume this is a technical detail we'll easily resolve once you're back from vacation; for now I believe this WIP PR to be a useful vehicle for me to publish and comment on my progress and for you to review it when you're back in office.

Thanks

Tomas

davidwrighton committed:
Autogenerated tests for context hopping in Virtual static interface methods

davidwrighton committed:
Adjust line endings

davidwrighton committed:
Refactor so that C# code is expressed as C# not as hand copied around IL

davidwrighton committed:
Fix new lines

davidwrighton committed:
Closer

davidwrighton committed:
Fixup expected strings

davidwrighton committed:
Move reporting to C# code

davidwrighton committed:
First pass at type hierarchy test

davidwrighton committed:
Fix the build

davidwrighton committed:
Add minimal covariant testing

davidwrighton committed:
Add minimal test for initial bringup

trylek committed:
Minimalistic JIT change to enable the new constrained. opcodes

trylek committed:
Initial runtime changes to start JITting Test_Scenario1

trylek committed:
MethodTable construction fixes letting me get to getCallInfo for the SVM

trylek committed:
Small improvements in TypeHierarchyTest_minimal

1) Make Statics a public class so that the IL can call into it;

2) Slightly modified the minimal test so that Method returns a
string that can be used for a minimalistic check of actual runtime
behavior;

3) Expanded the test to separately cover BaseScenario1 and
DerivedScenario1 as Test_Scenario1 and Test_Scenario2. Test_Scenario1
also had unbalanced stack (we never consumed the int32 return value
of Method) and the refactoring fixes that too.

Thanks

Tomas

trylek committed:
Initial implementation of SVM lookup; TypeHierarchyTest_minimal passes

I have cloned and slightly massaged the existing
TryResolveConstraintMethodApprox into a new method
TryResolveStaticVirtualConstraintMethodApprox for use when resolving
SVMs. I'm not yet 100% clear on some fine points in the implementation
but at the very least it's good enough to resolve the two SVMs used by
the TypeHierarchyTest_minimal smoke test where both scenarios pass
with this change.

Thanks

Tomas

trylek committed:
Move the SVM resolution variant out of the ValueType check

trylek committed:
Small improvements in the TypeHierarchy test generator

1) Fixed the 'interface' flag for the interface classes;

2) Hopefully fixed the override specifications in the derived methods;

3) Added a third scenario to the minimal test that poses interesting
challenges to the existing method table builder - the variation where
the BaseScenario class doesn't implement InterfaceScenario and only
the DerivedScenario class does by means of providing a MethodImpl
pointing at the base class - that one doesn't work yet.

4) For the complete TypeHierarchyTest suite, with these changes the
first scenario passes and the second one crashes the test with the issue
described under bullet (3). Just 373 more to go, non-zero pass rate is
always promising.

Thanks

Tomas

davidwrighton commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2943,10 +2946,6 @@ MethodTableBuilder::EnumerateClassMethods()
                 {
                     BuildMethodTableThrowException(BFA_VIRTUAL_PINVOKE_METHOD);
                 }
-                if(IsMdStatic(dwMemberAttrs))
```
We should not permit static virtual methods on classes/structs. By our ecma design, those should continue to throw.

davidwrighton commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -4770,17 +4769,15 @@ VOID MethodTableBuilder::TestMethodImpl(
         BuildMethodTableThrowException(IDS_CLASSLOAD_MI_MUSTBEVIRTUAL);
     }
     // Virtual methods cannot be static
-    if (IsMdStatic(dwDeclAttrs))
```
We should not permit static virtual methods on classes/structs. By our ecma design, those should continue to throw.

davidwrighton commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -5023,15 +5020,11 @@ MethodTableBuilder::ValidateMethods()
+                */
             }
         }
-
```
We should not permit static virtual methods on classes/structs. By our ecma design, those should continue to throw.

davidwrighton commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -5613,15 +5606,6 @@ MethodTableBuilder::ProcessMethodImpls()
-        // error before reaching this point.
-        CONSISTENCY_CHECK(!(!IsMdVirtual(it.Attrs()) && it.IsMethodImpl()));
-
-        if (!IsMdVirtual(it.Attrs()))
```
This should only be allowed for static methods

davidwrighton commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -2376,6 +2377,7 @@ MethodTableBuilder::EnumerateMethodImpls()
                 {
                     BuildMethodTableThrowException(IDS_CLASSLOAD_MI_ILLEGAL_TOKEN_BODY);
                 }
+                /* TODO: what to do about this check
```
The spec update describes the changes here. This check should be preserved for instance virtuals, but there should be a base check for static virtual overrides.

davidwrighton commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -2958,6 +2958,25 @@ ClassLoader::LoadApproxTypeThrowing(
         CorElementType type = ELEMENT_TYPE_END;
         IfFailThrowBF(sigptr.GetElemType(&type), BFA_BAD_SIGNATURE, pModule);
 
+        // TODO: For some reason, the base class BaseScenario1 - DerivedScenario1 inheritance
```
This is odd. Possibly we just have a ilasm bug.

davidwrighton commented on src/coreclr/vm/contractimpl.cpp:
```diff
@@ -434,9 +434,11 @@ DispatchMap::CreateEncodedMapping(
             curTargetSlot = -1;
             do
             {
+                /* TODO: right now it's unclear to me whether the SVMs should go in the method table or not, for now I assume they should.
```
My initial thought was to store this sort of detail in a parallel table to the existing scheme, but we'll want to take a close look here to see if this is the right approach.

davidwrighton commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1950,7 +1950,7 @@ extern "C" PCODE STDCALL PreStubWorker(TransitionBlock* pTransitionBlock, Method
 
         MethodTable* pDispatchingMT = NULL;
-        if (pMD->IsVtableMethod())
+        if (!pMD->IsStatic() && pMD->IsVtableMethod())
```
I don't like the idea that static methods would go in the vtable.

davidwrighton reviewed (changes requested):
I have a set of comments on the change in progress, although I'm really happy to see at least 1 test pass. Very exciting.

davidwrighton committed:
Fix issues discovered by Tomas
- Missing interface class flag on interfaces in TypeHierarchy tests
- Improper use of the class token when deriving, implementing interfaces or defining a MethodImpl
  - This caused incorrect behavior where ilasm will generate a typespec to a non-generic method instead of the correct typedef/typeref token
- Common C# code should be public
- Improved CheckForFailure code in type hierarchy tests

trylek commented:
Closing in favor of a new PR against David's private branch.
